dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/monitor/DatadogRumMonitor.kt

```diff
@@ -0,0 +1,165 @@
+                "Unable to stop view with key <$key> and class <$className>. " +
+                    "This can mean that the view was not started or already stopped."
+            )
+        } else if (startedEventData == null) {
+            devLogger.e(
+                "Unable to stop view with key <$key> and class <$className>. " +
+                    "The related data was inconsistent."
+            )
+        } else {
+            val updatedEvent = startedEvent.copy(
+                eventData = startedEventData.copy(
+                    durationNanoSeconds = System.nanoTime() - startedEventData.durationNanoSeconds,
+                    version = startedEventData.version + 1
+                ),
+                attributes = startedEvent.attributes.toMutableMap().apply {
+                    putAll(attributes)
+                }
+            )
+            writer.write(updatedEvent)
+        }
+        GlobalRumMonitor.updateViewId(null)
+    }
+
+    override fun startResource(key: Any, url: String, attributes: Map<String, Any?>) {
+        val eventData = RumEventData.Resource(
+            RumEventData.Resource.Kind.OTHER,
+            url,
+            System.nanoTime()
+        )
+        val event = RumEvent(
+            GlobalRumMonitor.getRumContext(),
+            timeProvider.getServerTimestamp(),
+            eventData,
+            attributes
+        )
+        activeResources[key] = event
+    }
+
+    override fun stopResource(key: Any, mimeType: String?, attributes: Map<String, Any?>) {
+        val startedEvent = activeResources.remove(key)
+        val startedEventData = startedEvent?.eventData as? RumEventData.Resource
+
+        if (startedEvent == null) {
+            devLogger.w(
+                "Unable to end resource with key <$key>. " +
+                    "This can mean that the resource was not started or already ended."
+            )
+        } else if (startedEventData == null) {
+            devLogger.e(
+                "Unable to end resource with key <$key>. The related data was inconsistent."
+            )
+        } else {
+            // TODO RUMM-248 detect Kind based on Content-Type
+            val updatedEvent = startedEvent.copy(
+                eventData = startedEventData.copy(
+                    durationNanoSeconds = System.nanoTime() - startedEventData.durationNanoSeconds
+                ),
+                attributes = startedEvent.attributes.toMutableMap().apply {
+                    putAll(attributes)
+                }
+            )
+            writer.write(updatedEvent)
+        }
+    }
+
+    override fun stopResourceWithError(key: Any, message: String, origin: String, throwable: Throwable) {
+        val startedEvent = activeResources.remove(key)
+        val startedEventData = startedEvent?.eventData as? RumEventData.Resource
+        val attributes = startedEvent?.attributes?.toMutableMap() ?: mutableMapOf()
+        attributes["http.url"] = startedEventData?.url
```

Comment:
it will be null otherwise ? is that ok on server side ?

Reply:
Updated